### PR TITLE
Refactor FXIOS-15851 WindowManager tabManager API should return an optional value

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-15851)

## :bulb: Description

Refactors the tabManager API in our `WindowManager` adopters to return an optional value. I've kept the Sentry logging in that func in place, so that unexpected or invalid UUIDs should still be visible to us.

Because we're still logging in that func, in most cases early returns are still reasonable and are preferable to us doing something like returning some other window's tabManager (which is very unsafe).

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If adding or modifying strings, I read the [guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/How-to-add-and-modify-Strings) and will request a string review from l10n
- [ ] If needed, I updated documentation and added comments to complex code

### CODE REVIEW

**Comment by mattreaganmozilla** [comment]
+cc @lmarceau Since this is ultimately more of a Tab change than an actual WindowManager change.

File: firefox-ios/Client/Application/WindowManager.swift [hunk 1]
```diff
@@ -73,11 +73,7 @@ protocol WindowManager {
     /// - Parameter tab: the UUID of the tab.
     /// - Returns: the UUID of the window hosting it (if available and open).
     @MainActor
-    func window(for tab: TabUUID) -> WindowUUID?
-
-    /// Convenience. Provides opportunity for safety checks or window validation.
-    @MainActor
-    func windowExists(uuid: WindowUUID) -> Bool
```
Comment (mattreaganmozilla) [inline firefox-ios/Client/Application/WindowManager.swift:80]: Removed `windowExists()` because I think it was a code smell. Generally callers should not need to directly check this.

File: firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift [hunk 2]
```diff
@@ -123,17 +121,12 @@ final class TabErrorTelemetryHelper {
         defaults.set(tabCounts, forKey: entryPoint.defaultsKey)
     }
 
-    /// It's possible for this telemetry helper to be called during onboarding flow before
```
Comment (lmarceau) [inline firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift:126]: That isn't a problem anymore (the comment) ?

Comment (mattreaganmozilla) [reply firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift:126]: The `tabManagerAvailable` func was added as a safety check, but with these refactors that safety check is now baked into the `tabManager(for:)` func, so each call site is forced to handle the potential nil value. So with this function gone this comment doesn't really have a great place to live now, since we're checking the optional value separately in each call site. We could maybe add the comment to the top of the file if you think it's still worth calling out?

Comment (lmarceau) [reply firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift:126]: Maybe we can do that yeah, I'll let you decide, you added that comment in the first place so you have more context than me! If you think it's useful to keep then lets move it, if not then 👋

**Review by lmarceau — APPROVED** [review]

File: firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift (hunk 2 above)
Comment (mattreaganmozilla) [reply firefox-ios/Client/Telemetry/TabErrorTelemetryHelper.swift:126]: Yeah this was a good call-out thank you; I've moved the comment to the top of the file.

File: firefox-ios/Client/Frontend/Browser/StartAtHome/StartAtHomeMiddleware.swift [hunk 3]
```diff
@@ -45,7 +45,7 @@ final class StartAtHomeMiddleware {
     /// - Returns: `true` if a homepage tab was selected and displayed, `false` otherwise.
     @MainActor
     private func startAtHomeCheck(windowUUID: WindowUUID) -> Bool {
-        let tabManager = windowManager.tabManager(for: windowUUID)
+        guard let tabManager = windowManager.tabManager(for: windowUUID) else { return false}
```
Comment (yoanarios) [inline firefox-ios/Client/Frontend/Browser/StartAtHome/StartAtHomeMiddleware.swift]: nit 
```suggestion
        guard let tabManager = windowManager.tabManager(for: windowUUID) else { return false }
```

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift [hunk 4]
```diff
@@ -37,7 +37,7 @@ final class MockWindowManager: WindowManager {
         wrappedManager.newBrowserWindowConfigured(windowInfo, uuid: uuid)
     }
 
-    func tabManager(for windowUUID: WindowUUID) -> TabManager {
```
Comment (yoanarios) [inline firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift:40]: Code wise changes seems right my only ask it will be to add support in the mock to return nil and add at least a test in TabManagerMiddlewareTests that tests the nil value instead of force unwrap to cover that case

**Review by yoanarios — COMMENTED** [review]
Changes look good just a small nit and unit test request

File: firefox-ios/Client/Frontend/Browser/StartAtHome/StartAtHomeMiddleware.swift (hunk 3 above)
Comment (mattreaganmozilla) [reply firefox-ios/Client/Frontend/Browser/StartAtHome/StartAtHomeMiddleware.swift]: Thank you, fixed

File: firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift (hunk 4 above)
Comment (mattreaganmozilla) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift:40]: I could be confused but isn't that a superfluous test? I don't think that actually exercises any real code in the client, right? LMK if I'm misunderstanding tho, I'm happy to update the unit tests.

Comment (yoanarios) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift:40]: Fair point, I agree we don’t want to add a test that only validates the mock. My concern is more about the new API contract: `tabManager(for:)` now returns an optional, and TabManagerMiddleware is one of the consumers that needs to behave safely when that value is nil.

Right now the production implementation explicitly returns nil for invalid/missing windows instead of falling back to an unsafe tab manager, so I think it’s worth having at least one unit test that exercises a real consumer path with a nil TabManager. That would protect us from future regressions where someone forces unwrapping or assumes the tab manager always exists again.

Comment (mattreaganmozilla) [reply firefox-ios/firefox-ios-tests/Tests/ClientTests/Mocks/MockWindowManager.swift:40]: I've pushed up unit tests in https://github.com/mozilla-mobile/firefox-ios/pull/33928/changes/35f40028b00f736e1e48856e68b5af6ddad79876 but LMK if you had something else specific in mind.

**Review by yoanarios — APPROVED** [review]
LGTM 🚀
